doc: Added a new page about nRF Connect Matter APIs (#14270)
nrfconnect/sdk-nrf: doc/nrf/protocols/matter/getting_started/matter_api.rst

There is no one unified, documented Matter SDK API that can be referenced when developing a custom application.
Instead, to learn how to interact with Matter library, a Matter firmware developer must follow exemplary source code of any of the existing Matter sample applications.
To aid that, Nordic provides unified API that wraps initialization of Matter specific components into more user friendly high level code.

Overall, the nRF Connect SDK Matter application code can be structured into the following steps:

1. Initialization of application specific components (e.g. initialization of hardware modules or proprietary Bluetooth LE services registration).
2. Initialization of Matter stack.
3. Start of the application main event loop.
4. Interaction between the application and Matter Data Model based on the Zigbee Cluster Library (ZCL) callbacks.

Points 1-3 can be implemented with the use of Nordic Matter API and utilities provided as a part of nRF Connect SDK Matter samples ``common`` modules (:file:`ncs/nrf/samples/matter/common`).
Point 4 requires the user to know details about ZCL callback functions that must be provided to interact with Matter Data Model.

> These callback are specific to the particular configuration of the Data Model (i.e. supported clusters) and therefore cannot be generalized or abstracted in a more user-friendly form.

Author: We miss plural form here, it should be "These callbacks". Will fix.

.. note::
   Before the interaction between the application and Matter Data Model is possible, the source files of the Data Model must be first generated with the `ZCL Advanced Platform`_ (ZAP tool).
   Please refer to the :ref:`ug_matter_gs_adding_cluster` for detailed description on how to configure and generate Matter Data model for your application.

nRF Connect Matter Initialization API

> *************************************

Author: LGTM, thanks.

The nRF Connect Matter API provides a set of functions and structures to initialize and manage a Matter server within a Matter application.

> This API allows developers to configure Matter stack components, event handlers, and networking backends.

Author: I think we miss something here: "server `to..."

   #ifdef CONFIG_CHIP_CRYPTO_PSA
     /** @brief Pointer to the user provided OperationalKeystore implementation. */
     chip::Crypto::OperationalKeystore *mOperationalKeyStore{ &sOperationalKeystoreDefault };
   #endif
     /** @brief Custom code to execute in the Matter main event loop before the server initialization. */
     CustomInit mPreServerInitClbk{ nullptr };
     /** @brief Custom code to execute in the Matter main event loop after the server initialization. */
     CustomInit mPostServerInitClbk{ nullptr };

     /** @brief Default implementation static objects that will be stripped by the compiler when above
      * pointers are overwritten by the application. */
   #ifdef CONFIG_CHIP_WIFI
     static chip::app::Clusters::NetworkCommissioning::Instance sWiFiCommissioningInstance;
   #endif
     static chip::CommonCaseDeviceServerInitParams sServerInitParamsDefault;
   #ifdef CONFIG_CHIP_FACTORY_DATA
     static chip::DeviceLayer::FactoryDataProvider<chip::DeviceLayer::InternalFlashFactoryData> sFactoryDataProviderDefault;
   #endif
   #ifdef CONFIG_CHIP_CRYPTO_PSA
     static chip::Crypto::PSAOperationalKeystore sOperationalKeystoreDefault;
   #endif
   };

> In this structure, each Matter component is defined as a public pointer to its generic interface that can be set with the concrete object.

Author: Yehh, but this is the wording developers usually use when talking about non-abstract objects that implement the generic interface.

And technically, it's only a *pointer* to the generic interface that can be set with concrete object. So we still need to combine both sentences to have "The generic interface pointer can be set with the concrete object." :)

> In this structure, each Matter component is defined as a public pointer to its generic interface that can be set with the concrete object.

Author: I think I will just remove this line. The merit is already conveyed in the next sentence.

> Moreover, each pointer is initially assigned with the default concrete implementation of the given interface that is stripped by the compiler in case the user overwrites it with customized implementation in the application.

Reviewer: ```suggestion
Each pointer is initially assigned with the default concrete implementation of the given interface.
The implementation is stripped by the compiler if the user overwrites it with a customized implementation in the application.
```
Is this still the correct meaning?

Author: Yes, I think it's fine. I will tune it a bit though.

> The nRF Connect Matter API consists of the following functions that can be used to initialize Matter components in proper order:

Reviewer: ```suggestion
The nRF Connect Matter API contains the following functions that can be used to initialize Matter components in proper order:
```

> * :c:func:`PrepareServer()`:
> 
> This function schedules initialization of Matter components including memory, server configuration and networking backend.
> Depending on the selected Kconfig options the initialization may also include factory data and operational key storage.
> Scheduling all initialization procedures to the Matter thread aims to provide a synchronization between all components and the application code.
> This function accepts :c:struct:`InitData` argument that contains implementation of all required Matter interfaces.
> If no argument is provided, this function uses the default-constructed :c:struct:`InitData` temporary object.
> After this function is used, the :c:func:`StartServer()` must be called to start the Matter thread, eventually execute the initialization and wait to synchronize the caller's thread with the Matter thread.

Reviewer: ```suggestion
:c:func:`PrepareServer()`
  This function schedules the initialization of Matter components, including memory, server configuration and networking backend.
  Depending on the selected Kconfig options, the initialization may also include factory data and operational key storage.
  All initialization procedures are scheduled to the Matter thread to provide a synchronization between all components and the application code.

  This function accepts an :c:struct:`InitData` argument that contains the implementation of all required Matter interfaces.
  If no argument is provided, this function uses the default-constructed :c:struct:`InitData` temporary object.
  After this function is used, the :c:func:`StartServer()` function must be called to start the Matter thread, eventually execute the initialization, and wait to synchronize the caller's thread with the Matter thread.
```

> * :c:func:`StartServer()`:
> 
> This is a blocking function that starts the Matter thread and waits until all Matter server components are initialized.

Reviewer: ```suggestion
:c:func:`StartServer()`:
  This is a blocking function that starts the Matter thread and waits until all Matter server components are initialized.
```

> * :c:func:`GetFactoryDataProvider()`:
> 
> This function returns generic pointer to the ``FactoryDataProvider`` object that was set during the initialization.
> It may come in handy when there is need to access factory data at the Matter server initialization stage or as a part of the post initialization callback (``mPostServerInitClbk`` in :c:struct:`InitData`).
> This function is available only if the :kconfig:option:`CONFIG_CHIP_FACTORY_DATA` Kconfig option is selected.

Reviewer: ```suggestion
:c:func:`GetFactoryDataProvider()`:
  This function returns the generic pointer to the ``FactoryDataProvider`` object that was set during the initialization.
  It can be used when you need to access factory data at the Matter server initialization stage or as a part of the post initialization callback (``mPostServerInitClbk`` in :c:struct:`InitData`).
  
  This function is only available if the :kconfig:option:`CONFIG_CHIP_FACTORY_DATA` Kconfig option is selected.
```

> For more details regarding nRF Connect Matter API please refer to the Doxygen commentary in the :file:`ncs/nrf/samples/matter/common/src/app/matter_init.h` header file.

Reviewer: ```suggestion
For more details regarding nRF Connect Matter initialization API, refer to the Doxygen commentary in the :file:`ncs/nrf/samples/matter/common/src/app/matter_init.h` header file.
```

> nRF Connect Matter Event Handler API
> ************************************

Reviewer: ```suggestion
Event handler API
*****************
```

> Matter SDK provides the notification scheme based on the public events that are propagated from the Matter stack to the application layer.

Reviewer: ```suggestion
The Matter SDK provides a notification scheme based on the public events that are propagated from the Matter stack to the application layer.
```

The nRF Connect Matter event handler API provides mechanisms to register and unregister custom functions that handle these events within an application.

> This module also includes a default handler that is used in nRF Connect SDK Matter samples and applications.
> The specific Matter events that can be handled in the application are listed in the :file:`ncs/modules/lib/matter/src/include/platform/CHIPDeviceEvent.h` file.
> The API consists of the following functions:

Reviewer: ```suggestion
This module also includes a default handler that is used in nRF Connect SDK Matter samples and applications.

The specific Matter events that can be handled in the application are listed in the :file:`ncs/modules/lib/matter/src/include/platform/CHIPDeviceEvent.h` header file.
The nRF Connect Matter API contains of the following functions that can be used to handle events:
```

> * :c:func:`RegisterEventHandler()`:
> 
> This function allows for registering the provided Matter event handler(``EventHandlerFunct``) in a thread safe manner.
> It is safe to call this function in the application after the Matter server was already initialized.

Reviewer: ```suggestion
:c:func:`RegisterEventHandler()`:
  This function is used to register the provided Matter event handler(``EventHandlerFunct``) in a thread-safe manner.
  It is safe to call this function in the application after the Matter server has already been initialized.
```

> * :c:func:`UnregisterEventHandler()`:
> 
> This function allows for unregistering the provided Matter event handler(``EventHandlerFunct``) in a thread safe manner.
> It is safe to call this function in the application after the Matter server was already initialized.

Reviewer: ```suggestion
:c:func:`UnregisterEventHandler()`:
  This function is used to unregister the provided Matter event handler(``EventHandlerFunct``) in a thread-safe manner.
  It is safe to call this function in the application after the Matter server has already been initialized.
```

> * :c:func:`DefaultEventHandler()`:
> 
> This is a nRF Connect Matter event handler function, that is registered in nRF Connect Matter Initialization API by default.
> This handler can be unregistered with :c:func:`UnregisterEventHandler()` in the application if needed.

Reviewer: ```suggestion
:c:func:`DefaultEventHandler()`:
  This is an nRF Connect Matter event handler function that is registered in the nRF Connect Matter Initialization API by default.
  You can unregister this handler with the :c:func:`UnregisterEventHandler()` function in the application if needed.
```

> For more details regarding nRF Connect Matter Event Handler API please refer to the Doxygen commentary in the :file:`ncs/nrf/samples/matter/common/src/app/matter_event_handler.h` header file.

Reviewer: ```suggestion
For more details regarding nRF Connect Matter event handler API, refer to the Doxygen commentary in the :file:`ncs/nrf/samples/matter/common/src/app/matter_event_handler.h` header file.
```

nRF Connect Matter API usage example

> ************************************

Reviewer: ```suggestion
====================================
```

> Combining both aforementioned nRF Connect Matter APIs the user can develop an application by following the initialization scheme listed below:

Reviewer: ```suggestion
Combining both aforementioned nRF Connect Matter APIs, you can develop an application by following the initialization scheme listed below:
```

.. code-block:: C++

   #include "app/matter_init.h"

   #include <zephyr/logging/log.h>

   LOG_MODULE_DECLARE(app, CONFIG_CHIP_APP_LOG_LEVEL);

   static void CustomMatterEventHandler(const ChipDeviceEvent *event, intptr_t /* unused */)
   {
     switch (event->Type) {
     case DeviceEventType::kCHIPoBLEAdvertisingChange:
     /* Custom code, e.g. control LED */
     break;
     }
   }

   CHIP_ERROR MatterAppInit()
   {
     /* Initialize Matter stack */
     ReturnErrorOnFailure(Nrf::Matter::PrepareServer(Nrf::Matter::InitData{ .mPostServerInitClbk = [] {
     LOG_INF("Matter server has been initialized.");
     return CHIP_NO_ERROR;
     } }));

     /* Register custom Matter event handler. */
     ReturnErrorOnFailure(Nrf::Matter::RegisterEventHandler(CustomMatterEventHandler, 0));

     /* Application specific initialization, e.g. hardware initialization.

     ...
     */

     /* Start Matter thread that will run the scheduled initialization procedure. */
     return Nrf::Matter::StartServer();
   }

> Note that the ``PrepareServer()`` call may contain more fields of the :c:struct:`InitData` being initialized.
> This function can be also called without any explicit argument. In that case, the default initialization will be provided.

Reviewer: ```suggestion
Note that the ``PrepareServer()`` call may contain more fields of the :c:struct:`InitData` being initialized, or can be called without any explicit argument.
If there is no explicit argument, the default initialization will be provided.
```

> For more references and examples on how to leverage nRF Connect Matter APIs please reach :ref:`Matter samples <matter_samples>`' source code in nRF Connect SDK.

Reviewer: ```suggestion
For more references and examples on how to leverage the nRF Connect Matter APIs, examine the source code for the :ref:`matter_samples` in the |NCS|.
```

Interacting with Matter Data Model
**********************************

> Matter SDK Data Model interaction with the user's code is based on callbacks that can be implemented by the application.

Reviewer: ```suggestion
The Matter SDK Data Model interacs with the user's code based on callbacks that can be implemented by the application.
```

> The generic callbacks that are common for Matter applications, regardless of the clusters configuration, are defined in :file:`ncs/modules/lib/matter/src/app/util/generic-callbacks.h` header file.

Reviewer: ```suggestion
The generic callbacks that are common for Matter applications, regardless of the clusters configuration, are defined in the :file:`ncs/modules/lib/matter/src/app/util/generic-callbacks.h` header file.
```

> The weak implementations of these functions, that can be overwritten in the application, are provided in :file:`ncs/modules/lib/matter/src/app/util/generic-callback-stubs.cpp` source file.

Reviewer: ```suggestion
The weak implementations of these functions, that can be overwritten in the application, are provided in the :file:`ncs/modules/lib/matter/src/app/util/generic-callback-stubs.cpp` source file.

```
Note empty line to create paragraph break.

For example, the :c:func:`MatterPostAttributeChangeCallback()` function is called by the Matter Data Model engine directly after an attribute value is changed.
The value passed into this callback is the value to which the attribute was set by the framework.

> Apart from the value, this function is called with the attribute path ( of ``chip::app::ConcreteAttributePath``type) that can be used to filter the cluster and particular attribute of interest.

Reviewer: ```suggestion
In addition to the value, this function is called with the attribute path ( of ``chip::app::ConcreteAttributePath`` type) that can be used to filter the cluster and particular attribute.
```

> :c:func:`MatterPostAttributeChangeCallback()` function is useful in case there is a need to provide the synchronization between the Data Model and the application state.

Reviewer: ```suggestion
The :c:func:`MatterPostAttributeChangeCallback()` function is useful if you need to provide the synchronization between the Data Model and the application state.
```

> For instance, the Matter device implementing a light bulb may drive the state of LED based on the ``On/Off`` attribute value.

Reviewer: ```suggestion
For instance, a Matter device that implements a light bulb may drive the state of the LED based on the ``On/Off`` attribute value.
```

> Every change of this attribute is reported by aforementioned callback and thus be captured in the application layer.

Reviewer: ```suggestion
Every change of this attribute is reported by the aforementioned callback and thus be captured in the application layer.

```
Note the added empty line.

Author: It should be "and thus can be ". Will fix.

> Apart from :c:func:`MatterPostAttributeChangeCallback()`, Matter defines other generic callbacks that can be employed in different use cases.

Reviewer: ```suggestion
In addition to the :c:func:`MatterPostAttributeChangeCallback()` function, Matter defines other generic callbacks that can be employed in different use cases.
```

> As an example, :c:func:`emberAfExternalAttributeReadCallback()` and :c:func:`emberAfExternalAttributeWriteCallback()` can be used to store and handle attributes externally, by bypassing the Matter Data Model framework.

Reviewer: ```suggestion
For example, the :c:func:`emberAfExternalAttributeReadCallback()` and :c:func:`emberAfExternalAttributeWriteCallback()` functions can be used to store and handle attributes externally, by bypassing the Matter Data Model framework.
```

> To learn the complete set of Matter generic callbacks please use the :file:`ncs/modules/lib/matter/src/app/util/generic-callbacks.h` header file and included Doxygen commentary a reference.

Reviewer: ```suggestion
To learn the complete set of Matter generic callbacks, refer to the :file:`ncs/modules/lib/matter/src/app/util/generic-callbacks.h` header file and included Doxygen commentary.
```

> The exemplary implementation of the :c:func:`MatterPostAttributeChangeCallback()` that can be used to control the DoorLock Matter device type is listed below:

Reviewer: ```suggestion
An example implementation of the :c:func:`MatterPostAttributeChangeCallback()` that can be used to control the Door Lock Matter device type is listed below:
```

.. code-block:: C++

   #include <app-common/zap-generated/ids/Clusters.h>
   #include <app/ConcreteAttributePath.h>
   #include <app/util/generic-callbacks.h>

   using namespace ::chip;
   using namespace ::chip::app::Clusters;

   void MatterPostAttributeChangeCallback(const chip::app::ConcreteAttributePath &attributePath, uint8_t type,
                   uint16_t size, uint8_t *value)
   {
     if (attributePath.mClusterId == DoorLock::Id &&
         attributePath.mAttributeId == DoorLock::Attributes::LockState::Id) {
       /* Post events only if current lock state is different than given */
       switch (*value) {
       case to_underlying(DlLockState::kLocked):
         /* Lock the physical door lock. */
         break;
       case to_underlying(DlLockState::kUnlocked):
       /* Unlock the physical door lock. */
         break;
       default:
         break;
       }
     }
   }

> Apart from Matter generic callbacks, Matter Data Model engine provides callbacks that are cluster specific.

Reviewer: ```suggestion
In addition to Matter generic callbacks, the Matter Data Model engine provides callbacks that are cluster-specific.
```

> These callbacks are usually defined was weak functions in the :file:`callback-stub.cpp` file that is generated together with other C++ source files when configuring clusters automatically with the `ZCL Advanced Platform`_ (ZAP tool).

Reviewer: ```suggestion
These callbacks are usually defined as weak functions in the :file:`callback-stub.cpp` file that is generated together with other C++ source files when configuring clusters automatically with the `ZCL Advanced Platform`_ (ZAP tool).
```

> In case of some clusters however, the different approach is used and related callbacks are defined within the source code that constitutes the implementation of the cluster itself.

Reviewer: ```suggestion
In case of some clusters, however, a different approach is used and related callbacks are defined within the source code that constitutes the implementation of the cluster itself.
```

> As an example, the DoorLock cluster server implementation defines application callbacks in :file:`ncs/modules/lib/matter/src/app/clusters/door-lock-server/door-lock-server.h` and places correlated weak implementations in :file:`ncs/modules/lib/matter/src/app/clusters/door-lock-server/door-lock-server-callback.cpp` file.

Reviewer: ```suggestion
As an example, the ``DoorLock`` cluster server implementation defines application callbacks in the :file:`ncs/modules/lib/matter/src/app/clusters/door-lock-server/door-lock-server.h` header file and places the related weak implementations in the :file:`ncs/modules/lib/matter/src/app/clusters/door-lock-server/door-lock-server-callback.cpp` file.
```

.. note::

>    Most of the Matter Data Model callback function names are prefixed with ``emberAf``. The reason for that is the fact that Matter Data Model inherits extensively from the Zigbee Ember Application Framework API.

Reviewer: ```suggestion
   Most of the Matter Data Model callback function names are prefixed with ``emberAf``.
   The reason for this is the fact that the Matter Data Model inherits extensively from the Zigbee Ember Application Framework API.
```

> In nRF Connect SDK, all Matter samples follow the same convention and implement described Matter Data Model callbacks in the :file:`zcl_callbacks.cpp` files which are populated as a part of the application source code.

Reviewer: ```suggestion
In the |NCS|, all Matter samples follow the same convention and implement the described Matter Data Model callbacks in the :file:`zcl_callbacks.cpp` files which are populated as a part of the application source code.
```

> You can review :file:`zcl_callbacks.cpp` of any nRF Connect SDK Matter sample to find exemplary implementations of various Data Model callbacks.

Reviewer: ```suggestion
You can review the :file:`zcl_callbacks.cpp` file of any |NCS| Matter sample to find example implementations of various Data Model callbacks.
```

> For instance, you can find the reference implementation of DoorLock specific Matter Data Model callbacks in :file:`ncs/nrf/samples/matter/lock/src/zcl_callbacks.cpp` source file.

Reviewer: ```suggestion
For instance, you can find the reference implementation of ``DoorLock``-specific Matter Data Model callbacks in the :file:`ncs/nrf/samples/matter/lock/src/zcl_callbacks.cpp` source file.
```

  This function is used to unregister the provided Matter event handler(``EventHandlerFunct``) in a thread-safe manner.
  It is safe to call this function in the application after the Matter server has already been initialized.

:c:func:`DefaultEventHandler()`:
  This is an nRF Connect Matter event handler function that is registered in the nRF Connect Matter Initialization API by default.
  You can unregister this handler with the :c:func:`UnregisterEventHandler()` function in the application if needed.

For more details regarding nRF Connect Matter event handler API, refer to the Doxygen commentary in the :file:`ncs/nrf/samples/matter/common/src/app/matter_event_handler.h` header file.

nRF Connect Matter API usage example
====================================

Combining both aforementioned nRF Connect Matter APIs, you can develop an application by following the initialization scheme listed below:

.. code-block:: C++

   #include "app/matter_init.h"

   #include <zephyr/logging/log.h>

   LOG_MODULE_DECLARE(app, CONFIG_CHIP_APP_LOG_LEVEL);

   static void CustomMatterEventHandler(const ChipDeviceEvent *event, intptr_t /* unused */)
   {
     switch (event->Type) {

>      case DeviceEventType::kCHIPoBLEAdvertisingChange:

Reviewer 2: nit: Is there any better event to showcase? Like CommissioningCompleted or sth? I mean i think it would the mean more for reader.

Author: Sure, we could use `PublicEventTypes::kCommissioningComplete`. I will create a followup task for that and spacing.

     /* Custom code, e.g. control LED */
     break;
     }
   }

   CHIP_ERROR MatterAppInit()
   {
     /* Initialize Matter stack */
     ReturnErrorOnFailure(Nrf::Matter::PrepareServer(Nrf::Matter::InitData{ .mPostServerInitClbk = [] {

>      LOG_INF("Matter server has been initialized.");

Reviewer 2: spaces here and below?

Btw. Not sure what is the standard for code snippets.. but only two spaces seems a bit narrow.

Author: @wiba-nordic do we have any standard here? Look like sth that could be changed indeed.

Reviewer: I don't think we actually have any standard for code snippets. The assumption is that they come from actual code, and would follow what is used there.
